c/src/ml-api-service-extension.c

```diff
@@ -0,0 +1,1122 @@
+{
+  ml_extension_s *ext = (ml_extension_s *) mls->priv;
+
+  /* Supposed internal function call to release handle. */
+  if (!ext)
+    return ML_ERROR_NONE;
+
+  /**
+   * Close message thread.
+   * If model inference is running, it may wait for the result in message thread.
+   * This takes time, so do not call join with extension lock.
+   */
+  ext->running = FALSE;
+  if (ext->msg_thread) {
+    g_thread_join (ext->msg_thread);
+    ext->msg_thread = NULL;
+  }
+
+  g_mutex_lock (&mls->lock);
+
+  /* Clear callback before closing internal handles. */
+  memset (&ext->callback, 0, sizeof (ml_service_callbacks_s));
+
+  if (ext->msg_queue) {
+    g_async_queue_unref (ext->msg_queue);
+    ext->msg_queue = NULL;
+  }
+
+  if (ext->conf_info) {
+    ml_option_destroy (ext->conf_info);
+    ext->conf_info = NULL;
+  }
+
+  if (ext->single) {
+    ml_single_close (ext->single);
+    ext->single = NULL;
+  }
+
+  if (ext->pipeline) {
+    ml_pipeline_stop (ext->pipeline);
+    ml_pipeline_destroy (ext->pipeline);
+    ext->pipeline = NULL;
+  }
+
+  if (ext->node_table) {
+    g_hash_table_destroy (ext->node_table);
+    ext->node_table = NULL;
+  }
+
+  g_mutex_unlock (&mls->lock);
+  g_free (ext);
+
+  return ML_ERROR_NONE;
+}
+
+/**
+ * @brief Creates a handle for machine learning service with configuration.
+ */
+int
+ml_service_new (const char *config_path, ml_service_h * handle)
+{
+  ml_service_s *mls;
+  ml_extension_s *ext;
+  g_autofree gchar *thread_name = g_strdup_printf ("ml-ext-msg-%d", getpid ());
+  int status;
+
+  check_feature_state (ML_FEATURE_SERVICE);
+
+  if (!handle) {
+    _ml_error_report_return (ML_ERROR_INVALID_PARAMETER,
+        "The parameter, 'handle' (ml_service_h), is NULL. It should be a valid pointer to create new instance.");
+  }
+
+  /* Init null. */
+  *handle = NULL;
+
+  mls = _ml_service_create_internal (ML_SERVICE_TYPE_EXTENSION);
+  if (mls == NULL) {
+    _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
+        "Failed to allocate memory for the ml-service handle. Out of memory?");
+  }
+
+  mls->priv = ext = g_try_new0 (ml_extension_s, 1);
+  if (ext == NULL) {
+    _ml_service_destroy_internal (mls);
+    _ml_error_report_return (ML_ERROR_OUT_OF_MEMORY,
+        "Failed to allocate memory for the service handle's private data. Out of memory?");
+  }
+
+  ext->type = ML_EXTENSION_TYPE_UNKNOWN;
+  ext->running = FALSE;
+  ext->timeout = DEFAULT_TIMEOUT;
+  ext->max_input = DEFAULT_MAX_INPUT;
+  ext->node_table = g_hash_table_new_full (g_str_hash, g_str_equal, g_free,
+      _ml_extension_node_info_free);
+
+  status = _ml_extension_conf_parse_json (mls, config_path);
+  if (status != ML_ERROR_NONE)
+    goto error;
+
+  g_mutex_lock (&mls->lock);
+
+  ext->msg_queue = g_async_queue_new_full (_ml_extension_msg_free);
+  ext->msg_thread = g_thread_new (thread_name, _ml_extension_msg_thread, mls);
+
+  /* Wait until the message thread has been initialized. */
+  g_cond_wait (&mls->cond, &mls->lock);
```

Comment:
How about checking  ```ext->running```?
```
if (!ext->running)
  g_cond_wait (&mls->cond, &mls->lock);
```

Reply:
The running flag is set true in msg-thread with lock and signal, so unnecessary here.

Reply:
Thank you.